auto_gptq/modeling/_utils.py

```diff
@@ -146,6 +147,51 @@ def make_quant(
             new_layer.device = ori_layer_device
             recurse_setattr(module, name, new_layer.to(ori_layer_device))
 
+def convert_gptq_v1_to_v2_format(
+    model,
+    v2_format: bool,
+    quantize_config: BaseQuantizeConfig,
+    qlinear_kernel: nn.Module,
+):
+    use_qigen = qlinear_kernel.QUANT_TYPE == "qigen"
+
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, qlinear_kernel):
+            if v2_format:
+                if use_qigen:
+                    submodule.zeros.data += 1
+                else:
+                    if quantize_config.bits == 2:
+                        submodule.qzeros.data += 0b01010101010101010101010101010101
+                    elif quantize_config.bits == 3:
+                        submodule.qzeros.data[:,range(0,submodule.qzeros.data.shape[1],3)] += 0b00100100100100100100100100100100
+                        submodule.qzeros.data[:,range(1,submodule.qzeros.data.shape[1],3)] += 0b10010010010010010010010010010010
+                        submodule.qzeros.data[:,range(2,submodule.qzeros.data.shape[1],3)] += 0b01001001001001001001001001001001
+                    elif quantize_config.bits == 4:
+                        submodule.qzeros.data += 0b00010001000100010001000100010001
+                    elif quantize_config.bits == 8:
+                        submodule.qzeros.data += 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+            else:
+                if use_qigen:
+                    submodule.zeros.data -= 1
+                else:
+                    if quantize_config.bits == 2:
+                        submodule.qzeros.data -= 0b01010101010101010101010101010101
+                    elif quantize_config.bits == 3:
+                        submodule.qzeros.data[:,range(0,submodule.qzeros.data.shape[1],3)] -= 0b00100100100100100100100100100100
+                        submodule.qzeros.data[:,range(1,submodule.qzeros.data.shape[1],3)] -= 0b10010010010010010010010010010010
+                        submodule.qzeros.data[:,range(2,submodule.qzeros.data.shape[1],3)] -= 0b01001001001001001001001001001001
+                    elif quantize_config.bits == 4:
+                        submodule.qzeros.data -= 0b00010001000100010001000100010001
+                    elif quantize_config.bits == 8:
+                        submodule.qzeros.data -= 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
```

Comment:
The real issue is v2 -> v1 that may underflow

The underflow is not addressed in the AutoGPTQ codebase https://github.com/AutoGPTQ/AutoGPTQ/blob/ea829c7bbe83561c2b1de26795b6592992373ef7/auto_gptq/nn_modules/qlinear/qlinear_cuda_old.py#L166 anyway, so I'm not sure

Reply:
@qwopqwop200  Can you comment on this underflow note as whether if it is an issue or not? Feel like this is the last blocking discussion.

Reply:
Almost all GPTQ checkpoints that exist today have sym=True (otherwise they won't work).
And the underflow issue is not a problem, at least when sym=True.
This is because when sym is True, the value of qzeros is fixed to 2** (bit-1).
Therefore, convert_gptq_v2_to_v1_format should only be applied to models converted from v1 to v2 or when sym=True.